src/displayapp/screens/Calculator.cpp

```diff
@@ -0,0 +1,350 @@
+}
+
+Calculator::~Calculator() {
+  lv_obj_clean(lv_scr_act());
+}
+
+static const char* buttonMap1[] = {"7", "8", "9", "/", "\n",
+                                   "4", "5", "6", "x", "\n",
+                                   "1", "2", "3", "-", "\n",
+                                   ".", "0", "=", "+", "",};
+
+static const char* buttonMap2[] = {"7", "8", "9", "(", "\n",
+                                   "4", "5", "6", ")", "\n",
+                                   "1", "2", "3", "^", "\n",
+                                   ".", "0", "=", "+", "",};
+
+Calculator::Calculator(DisplayApp* app, Controllers::MotorController& motorController)
+    : Screen(app), motorController{motorController} {
+  result = lv_label_create(lv_scr_act(), nullptr);
+  lv_label_set_long_mode(result, LV_LABEL_LONG_BREAK);
+  lv_label_set_text(result, "0");
+  lv_obj_set_size(result, 180, 60);
+  lv_obj_set_pos(result, 0, 0);
+  
+  returnButton = lv_btn_create(lv_scr_act(), nullptr);
+  lv_obj_set_size(returnButton, 52, 52);
+  lv_obj_set_pos(returnButton, 180, 8);
+  lv_obj_t* returnLabel;
+  returnLabel = lv_label_create(returnButton, nullptr);
+  lv_label_set_text(returnLabel, "<=");
+  lv_obj_align(returnLabel, nullptr, LV_ALIGN_CENTER, 0, 0);
+  returnButton->user_data = this;
+  lv_obj_set_event_cb(returnButton, eventHandler);
+  
+  buttonMatrix = lv_btnmatrix_create(lv_scr_act(), nullptr);
+  lv_btnmatrix_set_map(buttonMatrix, buttonMap1);
+  lv_obj_set_size(buttonMatrix, 240, 180);
+  lv_obj_set_pos(buttonMatrix, 0, 60);
+  buttonMatrix->user_data = this;
+  lv_obj_set_event_cb(buttonMatrix, eventHandler);
+  running = true;
+}
+
+void Calculator::eval() {
+  std::stack<char> input(std::deque<char>(0, ' '));
+  for (int8_t i = position - 1; i >= 0; i--) {
+    input.push(text[i]);
+  }
+  std::stack<CalcTreeNode*> output(std::deque<CalcTreeNode*>(0, nullptr));
+  std::stack<char> operators(std::deque<char>(0, ' '));
+  bool expectingNumber = true;
+  int8_t sign = +1;
+  while (!input.empty()) {
+    
+    if (isdigit(input.top())) {
+      char numberStr[31];
+      uint8_t strln = 0;
+      uint8_t pointpos = 0;
+      while (!input.empty() && (isdigit(input.top()) || input.top() == '.')) {
+        if (input.top() == '.') {
+          if (pointpos != 0) {
+            motorController.SetDuration(10);
+            return;
+          }
+          pointpos = strln;
+        } else {
+          numberStr[strln] = input.top();
+          strln++;
+        }
+        input.pop();
+      }
+      //replacement for strtod() since using that increased .txt by 76858 bzt
+      if (pointpos == 0) {
+        pointpos = strln;
+      }
+      double num = 0;
+      for (uint8_t i = 0; i <  pointpos; i++) {
+        num += (numberStr[i] - '0') * pow(10,   pointpos - i - 1 );
+      }
+      for (uint8_t i = 0; i < strln - pointpos; i++) {
+        num += (numberStr[i + pointpos] - '0') / pow(10,i+1);
+      }
+  
+      NumNode* number = new NumNode();
```

Comment:
I think there was some reason for using pointers. Ill have to take a look at that

Reply:
The reason was that you cant have the type of a field be abstract. I've now replaced the c-style pointers with shared_ptr (shared because there might be references in both the stack and individual nodes). Doing it this way appeases valgrind.

Reply:
Yes, I missed that these nodes get pushed onto the `output` stack.